packages/go_router/lib/src/delegate.dart

```diff
@@ -111,6 +117,51 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
     state.pop(result);
   }
 
+  /// Pop the Navigator's page stack until the predicate returns `true`.
+  void popUntil(PopUntilPredicate predicate) {
+    bool hasStoppedPopping = false;
+    bool popUntilPredicate(RouteMatchBase routeMatch, Route<dynamic> route) {
+      if (predicate(routeMatch, route)) {
+        hasStoppedPopping = true;
+        return true;
+      }
+      return false;
+    }
+
+    while (!hasStoppedPopping) {
+      final bool couldBePopped = _popWithPredicate(popUntilPredicate);
+      if (!couldBePopped) {
+        break;
+      }
+    }
+  }
+
+  bool _popWithPredicate(
+    PopUntilPredicate predicate,
+  ) {
+    NavigatorState? state;
+
+    if (navigatorKey.currentState?.canPop() ?? false) {
+      state = navigatorKey.currentState;
+    }
+    RouteMatchBase walker = currentConfiguration.matches.last;
+    while (walker is ShellRouteMatch) {
+      if (walker.navigatorKey.currentState?.canPop() ?? false) {
+        state = walker.navigatorKey.currentState;
+      }
+      walker = walker.matches.last;
+    }
+    int count = 0; // Only pop 1 page at the time.
+    state?.popUntil((Route<dynamic> route) {
+      if (count == 1 || predicate(walker, route)) {
+        return true;
+      }
+      count++;
+      return false;
+    });
+    return state != null;
+  }
```

Comment:
@chunhtai Do you have any feedback on this?

Reply:
is there a use case for `Route<dynamic> route`? it seems a bit out of place that we all deal with RouteBase and GoRouterState in this package.

Reply:
The case I'm thinking about and I've been facing in some applications is when there are dialogs / bottom sheets / popup menus (or any other imperative routes) open



I also worked on some projects where the code base uses both `go_router` and `Navigator.of(context).push`. Either because they didn't get the time to migrate all the pages yet or because they don't want those pages to be linked to a specific URL on the web

Reply:
I see, my suggestion would be changing routematch to gorouterstate.